ledger/complete/wal/checkpoint_v6_writer.go

```diff
@@ -0,0 +1,559 @@
+		// (from level 0 to subtrieLevel) using topLevelNodes,
+		// node iterator skips subtrie as visited nodes.
+		subtrieRootNodes[root] = traversedSubtrieNodes[root]
+	}
+
+	// -1 to account for 0 node meaning nil
+	totalNodeCount := nodeCounter - 1
+
+	// write total number of node as footer
+	footer := encodeSubtrieFooter(totalNodeCount)
+	_, err = writer.Write(footer)
+	if err != nil {
+		return nil, 0, 0, fmt.Errorf("cannot write checkpoint subtrie footer: %w", err)
+	}
+
+	// write checksum to the end of the file
+	crc32Sum := writer.Crc32()
+	_, err = writer.Write(encodeCRC32Sum(crc32Sum))
+	if err != nil {
+		return nil, 0, 0, fmt.Errorf("cannot write CRC32 checksum %v", err)
+	}
+
+	return subtrieRootNodes, totalNodeCount, crc32Sum, nil
+}
+
+// storeTopLevelNodes stores the top level nodes into a checkpoint part file.
+// it returns (the index of root nodes, the total number of top level nodes)
+// any error returned are exceptions
+func storeTopLevelNodes(
+	tries []*trie.MTrie,
+	subTrieRootIndices map[*node.Node]uint64,
+	initNodeCounter uint64,
+	writer io.Writer) (
+	map[*node.Node]uint64,
+	uint64,
+	error) {
+	scratch := make([]byte, 1024*4)
+	nodeCounter := initNodeCounter
+	var err error
+	for _, t := range tries {
+		root := t.RootNode()
+		if root == nil {
+			continue
+		}
+		// if we iterate through the root trie with an empty visited nodes map, then it will iterate through
+		// all nodes at all levels. In order to skip the nodes above subtrieLevel, since they have been seralized in step 1,
+		// we will need to pass in a visited nodes map that contains all the subtrie root nodes, which is the topLevelNodes.
+		// The topLevelNodes was built in step 1, when seralizing each subtrie root nodes.
+		nodeCounter, err = storeUniqueNodes(root, subTrieRootIndices, nodeCounter, scratch, writer, func(uint64) {})
+		if err != nil {
+			return nil, 0, fmt.Errorf("fail to store nodes in step 2 for root trie %v: %w", root.Hash(), err)
+		}
+	}
+
+	topLevelNodesCount := nodeCounter - initNodeCounter
+	return subTrieRootIndices, topLevelNodesCount, nil
+}
+
+// storeRootNodes stores the hash and node index of root nodes into a checkpoint part file.
+// any error returned are exceptions
+func storeRootNodes(
+	tries []*trie.MTrie,
+	topLevelNodes map[*node.Node]uint64,
+	encodeNode NodeEncoder,
+	writer io.Writer) error {
+	scratch := make([]byte, 1024*4)
+	for _, t := range tries {
+		rootNode := t.RootNode()
+
+		// Get root node index
+		rootIndex, found := topLevelNodes[rootNode]
+		if !found {
+			rootHash := t.RootHash()
+			return fmt.Errorf("internal error: missing node with hash %s", hex.EncodeToString(rootHash[:]))
+		}
+
+		encTrie := encodeNode(t, rootIndex, scratch)
+		_, err := writer.Write(encTrie)
+		if err != nil {
+			return fmt.Errorf("cannot serialize trie: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func storeFooter(topLevelNodesCount uint64, rootTrieCount uint16, writer io.Writer) error {
```

Comment:
encoding the header using proto or json (or whatever std we use here) instead of hand packing the bytes.

also, can you structure them as objects instead of global functions?

Reply:
>encoding the header using proto or json (or whatever std we use here) instead of hand packing the bytes.

The header has always been hand packed bytes, so that we can read the same format and recognize the version number encoded in it. And using the same format allows me to do round trip testing to confirm converting to v6 then converting back to v5 would produce the same result as converting to v5 directly. The round trip testing helps verifying the v6 algorithm 

>also, can you structure them as objects instead of global functions?
with functions, it's clear to see what it depends on, easy to write tests. Not sure what's the benefit of using objects.

Reply:
i'm not against hand packing the version (and the header content size), but the header content does not need to be hand packed.

Reply:
Footer is hand packed using fix sized so it can always be read at same offset from EOF.  Using a codec can make the offset from EOF a nondeterministic distance (depending on the format and/or codec settings used).

Footer contains data that need to be read before processing main file content.

I think subtrie file header is hand packed because it just contains magic bytes + version.  I have some thoughts about summary file but that's a separate topic from footers and headers.